bftengine/src/bftengine/ControlStateManager.cpp

```diff
@@ -87,4 +90,61 @@ void ControlStateManager::clearCheckpointToStopAt() {
   saveReservedPage(0, data.size(), data.data());
 }
 
+std::pair<bool, std::string> ControlStateManager::canUnwedge() {
+  std::optional<std::string> result;
+  if (!bftEngine::IControlHandler::instance()->isOnNOutOfNCheckpoint()) {
+    return {false, "Replica has not reached the wedge point yet"};
+  }
+
+  std::string replica_id = std::to_string(ReplicaConfig::instance().getreplicaId());
+  auto sigManager = impl::SigManager::instance();
+  std::string sig(sigManager->getMySigLength(), '\0');
+  sigManager->sign(replica_id.c_str(), replica_id.size(), sig.data(), sig.size());
+
+  return {true, sig};
+}
+
+bool ControlStateManager::verifyUnwedgeSignatures(
+    std::vector<std::pair<uint64_t, std::vector<uint8_t>>> const& signatures) {
+  auto sigManager = impl::SigManager::instance();
+
+  size_t quorum = ReplicaConfig::instance().numReplicas;
+  if (signatures.size() < quorum) {
+    LOG_INFO(GL, "Not enough signatures for verification");
+    return false;
+  }
+  size_t verified_sigs = 0;
+  for (auto const& sig : signatures) {
+    // SigManager cannot verify signature coming from same replica
+    // Generate the signature again and compare them.
+    std::string data = std::to_string(sig.first);
+    std::string signature(sig.second.begin(), sig.second.end());
+    bool valid = false;
+    if (sig.first == ReplicaConfig::instance().replicaId) {
+      verified_sigs++;
+      std::string sign(sigManager->getMySigLength(), '\0');
```

Comment:
Why do you need to sign every time again, instead of keeping the signature and generate it once on ctor?
Signing is expansive.

Reply:
This comment is valid only if you decided to sign the replica id over and over without "salt"